zluda/src/impl/context.rs

```diff
@@ -1,23 +1,86 @@
-use super::{driver, FromCuda, ZludaObject};
+use super::{FromCuda, ZludaObject, module};
 use cuda_types::cuda::*;
 use hip_runtime_sys::*;
-use rustc_hash::FxHashSet;
-use std::{cell::RefCell, ptr, sync::Mutex};
+use rustc_hash::{FxHashSet, FxHashMap};
+use std::{cell::RefCell, ptr, sync::Mutex, ffi::c_void};
+
 
 thread_local! {
-    pub(crate) static CONTEXT_STACK: RefCell<Vec<(CUcontext, hipDevice_t)>> = RefCell::new(Vec::new());
+    pub(crate) static STACK: RefCell<Vec<(CUcontext, hipDevice_t)>> = RefCell::new(Vec::new());
 }
 
 pub(crate) struct Context {
     pub(crate) device: hipDevice_t,
-    pub(crate) mutable: Mutex<OwnedByContext>,
+    pub(crate) state: Mutex<ContextState>,
+}
+
+pub(crate) struct ContextState {
+    pub(crate) ref_count: u32,
+    pub(crate) flags: u32,
+    pub(crate) modules: FxHashSet<CUmodule>,
+    pub(crate) storage: FxHashMap<usize, StorageData>,
 }
 
-pub(crate) struct OwnedByContext {
-    pub(crate) ref_count: usize, // only used by primary context
-    pub(crate) _memory: FxHashSet<hipDeviceptr_t>,
-    pub(crate) _streams: FxHashSet<hipStream_t>,
-    pub(crate) _modules: FxHashSet<CUmodule>,
+pub(crate) struct StorageData {
+    pub(crate) value: usize,
+    pub(crate) reset_cb: Option<extern "system" fn(CUcontext, *mut c_void, *mut c_void)>,
+    pub(crate) handle: CUcontext,
+}
+
+impl ContextState {
+    pub(crate) fn new() -> Self {
+        ContextState {
+            ref_count: 0,
+            flags: 0,
+            modules: FxHashSet::default(),
+            storage: FxHashMap::default(),
+        }
+    }
+
+    pub(crate) fn reset(&mut self) -> CUresult {
+        for (key, data) in self.storage.iter_mut() {
+            if let Some(_cb) = data.reset_cb {
+                // TODO: check that these callbacks do not call into the CUDA driver
+                _cb(data.handle, *key as *mut c_void, data.value as *mut c_void);
+            }
+        }
+        self.ref_count = 0;
+        self.flags = 0;
+        for hmod in self.modules.drain() {
+            super::drop_checked::<module::Module>(hmod)?;
```

Comment:
This is still a problem: first failing module will prevent all other drop_checked to run. It should either be a `.drain().fold()` that folds error codes or something along the lines:
```
let mut result = Ok(());
for hmod in self.modules.drain() {
    result = combine_errors(result, super::drop_checked::<module::Module>(hmod));
}
return result;
```